File: android-client/app/src/main/java/com/example/azurecommunication/data/repository/AcsRepository.kt
package com.example.azurecommunication.data.repository

import com.example.azurecommunication.data.api.AcsApiService
import com.example.azurecommunication.data.models.AcsUser
import com.example.azurecommunication.data.models.CreateThreadRequest
import com.example.azurecommunication.data.models.JoinThreadRequest
import com.example.azurecommunication.data.models.TokenRequest
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AcsRepository @Inject constructor(
    private val api: AcsApiService
) {
    suspend fun getAcsToken(displayName: String): Result<AcsUser> {
        return runCatching {
            val response = api.createToken(TokenRequest(displayName))
            AcsUser(
                userId = response.userId,
                token = response.token,
                displayName = displayName
            )
        }
    }

    suspend fun createChatThread(
        userId: String,
        displayName: String,
        topic: String = "Chat"
    ): Result<String> {
        return runCatching {
            val response = api.createChatThread(
                CreateThreadRequest(
                    userId = userId,
                    displayName = displayName,
                    topic = topic
                )
            )
            response.threadId
        }
    }

    suspend fun joinChatThread(
        threadId: String,
        userId: String,
        displayName: String
    ): Result<Unit> {
        return runCatching {
            api.joinChatThread(
                threadId = threadId,
                request = JoinThreadRequest(
                    userId = userId,
                    displayName = displayName
                )
            )
            Unit
        }
    }
}

File: ios-client/AzureCommunication/AzureCommunication/Data/Repository/AcsRepository.swift
//
//  AcsRepository.swift
//  AzureCommunication
//

import Foundation

actor AcsRepository {
    static let shared = AcsRepository()

    private let apiService = AcsApiService.shared

    private init() {}

    func getAcsToken(displayName: String) async throws -> AcsUser {
        let response = try await apiService.createToken(displayName: displayName)
        return AcsUser(
            userId: response.userId,
            token: response.token,
            displayName: displayName
        )
    }

    func createChatThread(userId: String, displayName: String, topic: String) async throws -> String {
        let response = try await apiService.createChatThread(
            userId: userId,
            displayName: displayName,
            topic: topic
        )
        return response.threadId
    }

    func joinChatThread(threadId: String, userId: String, displayName: String) async throws {
        _ = try await apiService.joinChatThread(
            threadId: threadId,
            userId: userId,
            displayName: displayName
        )
    }
}
